### PR TITLE
`bundle validate`: enable optional validators

### PR DESCRIPTION
**Pull request by estroz**
**Description of the change:**
* internal/cmd/operator-sdk/bundle: move validate-related functionality to the validate package, add `--select-optional` and `--list-optional` to select and list optional validators to run

**Motivation for the change:** optional validators like the [OperatorHub.io validator]() should be available but not run by default in `bundle validate`. New optional validators can be added and reviewed in a PR.

/cc @dinhxuanvu @joelanford @shawn-hurley 

/kind feature

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [x] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/cmd/operator-sdk/bundle/validate/optional.go
Comment (estroz) [inline internal/cmd/operator-sdk/bundle/validate/optional.go:40]: It might be nice to add a `GetName()` method to the `Validator` interface for error reporting and selection @kevinrizza @dinhxuanvu.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:36]: IMO we ought to use the website links.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:53]: I do not think that we should adopt the nomenclature Namespace to define a registry. it shows lead to misunderstanding.  Also, I do not think that we should add the export instead we should do as it is done in kb in our docs and other places here now. 

> mycommand  IMG=<some-registry>/<project-name>:<tag>

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:60]:
```suggestion
  # Use the built-in Makefile targets to build and push your operator. Make sure to define IMG when you call make:
  $ make docker-build docker-push IMG=<some-registry>/<project-name>:<tag>

  **NOTE**: To allow the cluster pull the image the repository needs to be set as public or you must configure an image pull secret.
  
  # Ensure the image with modified metadata and Dockerfile is valid.
  $ operator-sdk bundle validate <some-registry>/<project-name>:<tag>
```

I do not think that we should adopt the nomenclature Namespace to define a registry. it shows lead to misunderstanding.  Also, I do not think that we should add the export instead we should do as it is done in kb in our docs and other places here now. 

> `mycommand  IMG=<some-registry>/<project-name>:<tag>`


Also, IMO we should tell for the users use the make targets which reduces the steps and make its easier.

Comment (camilamacedo86) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:60]: Also, leI think we need to try to keep a consistency with the doc as well. See; https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:68]: Should not the second command here has an output as well?

File: internal/cmd/operator-sdk/bundle/validate/optional.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/optional.go:50]: WDYT about we try to persuade here the DDD principle?
See that we can have a BundleValidationOptional structure and then:

(b *BundleValidationOptional) run()
(b *BundleValidationOptional) list()

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (jmrodri) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:36]: @camilamacedo86 I don't see this design doc on the website anywhere. If there is a good description about them on the site, I will agree.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (jmrodri) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:68]: Agreed I'd like to see some of the output at least.

File: internal/cmd/operator-sdk/bundle/validate/optional.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/optional.go:50]: These functions are wrappers for the exact design you suggest (methods on `validators`).

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:36]: There isn't a non-Github hosted link.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:60]: The image tag format goes: `<registry host>/<registry namespace>/<repository name>:<tag>`. I don't see how using different names for something well understood helps users. The docs clearly state "registry namespace" too, so its pretty clear what this refers to.

Will update these docs to use "memcached-operator".

Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go]: Also updated `NAMESPACE -> REG_NAMESPACE`.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go]: Done.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:53]: IHMO we ought add here the same quickstart steps. 
If in each place we create a new nomenclature it might start to be confusing. 
Also, in POV `NAMESPACE` for your registry can lead to misunderstanding.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go]: IHMO we ought add here the same quickstart steps. 
If in each place we create a new nomenclature it might start to be confusing. 
Also, in POV `NAMESPACE` for your registry can lead to misunderstanding.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:57]:
```suggestion
  # Use the built-in Makefile targets to build and push your operator. Make sure to define IMG when you call make:
  
  $ make docker-build docker-push IMG=<some-registry>/<project-name>:<tag>
  
  **NOTE**: To allow the cluster pull the image the repository needs to be set as public or you must configure an image pull secret.
```

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:60]: why we need to call the bundle validate if it is called in the make bundle?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:48]: Why it is required if it is called in the make bundle already?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:56]: why not use the make docker-bundle instead of it?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:57]: why not use the makefile target scaffolded instead here?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:56]: IMO we should not add the output.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:69]: Ditto above.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:82]: Could we not explain and describe the command without adding its output?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:56]: Did you not [ask](https://github.com/operator-framework/operator-sdk/pull/3719#discussion_r470592361) for output?

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:59]:
```suggestion
  $ operator-sdk bundle validate <some-registry>/<project-name>:<tag>
```

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:56]:
```suggestion
  $ make bundle-build BUNLDE_IMG=<some-registry>/<project-name>:<tag>
  $ make docker-push IMG=<some-registry>/<project-name>:<tag>
```

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:59]:
```suggestion
  $ operator-sdk bundle validate <some-registry>/<project-name>:<tag>
```

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (camilamacedo86) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:52]:
```suggestion
NOTE: To allow the cluster pull the image the repository needs to be set as public or you must configure an image pull secret.
```

**Review by camilamacedo86 — COMMENTED** [review]
IMO: we just need to update the command description and re-generate the docs that will use it to be in the same way that we did for the quickstart. Then, after that, it is /lgtm for me too. See my comments.

**Review by camilamacedo86 — APPROVED** [review]
/lgtm after address the nit. Keep the cmd description following the same approach and nomenclatures adopted in https://sdk.operatorframework.io/docs/building-operators/golang/quickstart/

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:59]: This command can be used outside of a project.

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (estroz) [reply internal/cmd/operator-sdk/bundle/validate/cmd.go:55]: I've added the explicit requirement that the image be pullable to the example.

**Review by camilamacedo86 — APPROVED** [review]
I am ok with 👍 
/lgtm 

Regards the flag names are nice to check with at least more one person since will be harder we are able to remove/change they after.

File: changelog/fragments/bundle-validate-optional-validators.yaml
Comment (camilamacedo86) [inline changelog/fragments/bundle-validate-optional-validators.yaml:11]: HI @estroz, 

Would be nice add the checks for this feature in the e2e tests for all types since it would ensure the quality of the provided features with Helm/Ansible and GO. Could we do that?

File: changelog/fragments/bundle-validate-optional-validators.yaml
Comment (camilamacedo86) [inline changelog/fragments/bundle-validate-optional-validators.yaml:7]: @estroz, 

Scorecard has a similar feature. Should not better we use the same names? 
--list-selector an --selector? 

Since we cannot remove/change the flags I think it is very important we be very conscious over its names. 
/hold

Comment (estroz) [reply changelog/fragments/bundle-validate-optional-validators.yaml:7]: There are always required validators being run, so you're only able to select optional validators, therefore `--{list-}selector` would be disingenuous and `--{list,select}-optional` makes sense.

Comment (camilamacedo86) [reply changelog/fragments/bundle-validate-optional-validators.yaml:7]: OK. Far enough.

**Comment by camilamacedo86** [comment]
HI @estroz, 

It. is missing rebase with the master and add the checks in the e2e tests. Otherwise, it shows well to move forward. 
Great job 👍

**Pull request merged**
